bundles/io/org.eclipse.smarthome.io.json.gson/src/main/java/org/eclipse/smarthome/io/json/gson/internal/InternalGsonBindingService.java

```diff
@@ -0,0 +1,78 @@
+/**
+ * Copyright (c) 2014,2018 Contributors to the Eclipse Foundation
+ *
+ * See the NOTICE file(s) distributed with this work for additional
+ * information regarding copyright ownership.
+ *
+ * This program and the accompanying materials are made available under the
+ * terms of the Eclipse Public License 2.0 which is available at
+ * http://www.eclipse.org/legal/epl-2.0
+ *
+ * SPDX-License-Identifier: EPL-2.0
+ */
+package org.eclipse.smarthome.io.json.gson.internal;
+
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Map;
+
+import org.eclipse.smarthome.io.json.JsonBindingService;
+import org.eclipse.smarthome.io.json.gson.GsonBindingService;
+import org.eclipse.smarthome.io.json.gson.GsonTypeAdapterProvider;
+import org.osgi.service.component.annotations.Activate;
+import org.osgi.service.component.annotations.Component;
+import org.osgi.service.component.annotations.Reference;
+import org.osgi.service.component.annotations.ReferenceCardinality;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * JSON data binding implementation that uses Google Gson to serialize and deserialize between
+ * Java objects and JSON.
+ *
+ * @author Flavio Costa - Initial contribution and API
+ *
+ * @param <T> Type to be serialized or deserialized.
+ */
+@Component(immediate = true, service = JsonBindingService.class, property = {
+        InternalGsonBindingService.FORMAT_OUTPUT_PROPERTY + ":Boolean=false" })
+public class InternalGsonBindingService<T> extends GsonBindingService<T> {
+
+    /**
+     * Logger for this class.
+     */
+    private final Logger logger = LoggerFactory.getLogger(InternalGsonBindingService.class);
+
+    /**
+     * Component property that determines whether JSON output should be formatted (easier to read) or not (less data to
+     * be transmitted).
+     */
+    static final String FORMAT_OUTPUT_PROPERTY = "formatted-output";
+
+    /**
+     * Providers for Gson type adapters.
+     */
+    private final List<GsonTypeAdapterProvider> providers = new ArrayList<>();
+
+    @Reference(cardinality = ReferenceCardinality.MULTIPLE)
+    public void addTypeAdapterProvider(GsonTypeAdapterProvider provider) {
+        providers.add(provider);
+    }
+
+    public void removeTypeAdapterProvider(GsonTypeAdapterProvider provider) {
+        providers.remove(provider);
+    }
+
+    @Activate
+    public void activate(Map<String, Object> properties) {
+        Boolean prettyPrinting = (Boolean) properties.get(FORMAT_OUTPUT_PROPERTY);
```

Comment:
Originally I was having this on property files, but it sound overkill and it's something typically changed only for temporary debug purposes. It is now set on the `@Component` annotation on the class itself, so it is not something that would easily go wrong... but it would actually be nicer to change this to do the "formatted output" if log levels are set to DEBUG, but keep the unformatted Json output for INFO or less detailed logging levels. Do you know if this is something we can get from Logger or from somewhere else?

Reply:
How about using `logger.isDebugEnabled()`?